File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.java
class Solution {
    public int search(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;

        while(low < high) {
            int mid = (low + high) / 2;

            if(nums[mid] > nums[high]) {
                low = mid + 1;
            }
            else {
                high = mid;
            }
        }

        int pivot = low;

        int result = binarySearch(nums, target, 0, pivot - 1);

        if(result != -1) {
            return result;
        }
        return binarySearch(nums, target, pivot, nums.length - 1);
    }
    public int binarySearch(int[] nums, int target, int left, int right) {
        while(left <= right) {
            int mid = (left + right) / 2;
            if(nums[mid] == target) {
                return mid;
            }
            else if(nums[mid] < target) {
                left = mid + 1;
            }
            else {
                right = mid - 1;
            }
        }
        return -1;
    }
}

File: Data Structures & Algorithms/find-first-and-last-position-of-element-in-sorted-array/submission-0.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        leftval = self.binarySearch(nums, target, True)
        rightval = self.binarySearch(nums, target, False)
        return [leftval,rightval]

        
    def binarySearch(self, nums, target, leftBias) -> int:
        #implement the binary search here
        lo = 0
        hi = len(nums) -1 
        index = -1

        while lo <= hi:
            mid = lo + (hi - lo) // 2

            #check the right side
            if target > nums[mid]:
                lo = mid + 1
            elif target < nums[mid]:
                hi = mid - 1
            else:
                index = mid
                if leftBias:
                    hi = mid - 1
                else:
                    lo = mid + 1
            
        return index